File: c/data_structures/linked_list/singlyLinkedList.c
// This implementation has both HEAD and TAIL pointers.
// As a fun exercise, the HEAD and TAIL are passed to the functions as pointers to the pointers.
//
#include <stdio.h>
#include <stdlib.h>

struct node {
    struct node *next;
    int data;
};

struct node *addNode(struct node **head, struct node **tail, int data) {
    struct node *newNode;

    if (!(newNode = malloc(sizeof(struct node)))) {
        fprintf(stderr, "Could not allocate memory for new node!");
        exit(1);
    }

    if (*head == NULL) {
        newNode->data = data;
        newNode->next = NULL;
        *head = newNode;
        *tail = newNode;
    } else {
        struct node *curr = *head;

        while (curr->next) {
            curr = curr->next;
        }

        curr->next = newNode;
        newNode->data = data;
        newNode->next = NULL;

        if (newNode->next == NULL) {
            *tail = newNode;
        }
    }

    return newNode;
}

struct node *insertAfter(struct node **head, struct node **tail, struct node *elem, int data) {
    struct node *newNode, *curPos = *head;

    if (!(newNode = malloc(sizeof(struct node)))) {
        fprintf(stderr, "Could not allocate memory for new node!");
        exit(1);
    }

    if (elem == NULL) {
        if (*head == NULL) {
            newNode->data = data;
            newNode->next = NULL;

            *head = newNode;
            *tail = newNode;

            return newNode;
        } else {
            newNode->data = data;
            newNode->next = *head;

            *head = newNode;

            if (*tail == NULL) {
                *tail = newNode;
            }

            return newNode;
        }
    }

    while (curPos) {
        if (curPos == elem) {
            newNode->data = data;
            newNode->next = curPos->next;
            curPos->next = newNode;

            // Update TAIL pointer if the new node is the last.
            if (newNode->next == NULL) {
                *tail = newNode;
            }

            return newNode;
        }

        curPos = curPos->next;
    }

    // Insert position not found, free newNode memory!
    free(newNode);

    return NULL;
}

int removeHead(struct node **head, struct node **tail) {
    struct node *newHead;

    if (head && *head) {
        newHead = (*head)->next;
        free(*head);
        *head = newHead;

        if (*head == NULL) {
            *tail = NULL;
        }
    }
}

int removeNode(struct node **head, struct node **tail, struct node *toRemove) {
    struct node *curPos = *head;

    if (toRemove == NULL) {
        return 1;
    }

    if (*head == toRemove) {
        *head = toRemove->next;
        free(toRemove);

        // Special case for one element list.
        if (*head == NULL) {
            *tail == NULL;
        }

        return 0;
    }

    while (curPos) {
        if (curPos->next == toRemove) {
            curPos->next = toRemove->next;
            free(toRemove);

            // Update TAIL pointer when deleting the last element.
            if (curPos->next == NULL) {
                *tail = curPos;
            }

            return 0;
        }

        curPos = curPos->next;
    }

    return 1;
}

void removeList(struct node **head, struct node **tail) {
    struct node *toDelete = *head;

    while (toDelete) {
        struct node *next = toDelete->next;
        free(toDelete);
        toDelete = next;
    }

    *head = NULL;
    *tail = NULL;
}

void main(void) {
    struct node *head = NULL;
    struct node *tail = NULL;

    struct node *link1= addNode(&head, &tail, 1);
    struct node *link2 = addNode(&head, &tail, 2);
//     struct node *link3 = addNode(&head, &tail, 4);
//     struct node *link4 = addNode(&head, &tail, 8);
//     struct node *link5 = addNode(&head, &tail, 16);
//     struct node *link6 = addNode(&head, &tail, 32);
//     struct node *link7 = addNode(&head, &tail, 64);

    printf("%d\n", head->data);
    printf("%d\n", tail->data);
    removeHead(&head, &tail);
    printf("%d\n", head->data);
    printf("%d\n", tail->data);
    removeHead(&head, &tail);
    printf("%d\n", head);
    printf("%d\n", tail);
}

File: python/data_structures/linked_list/singlyLinkedList.py
class Node:
    ''' represents each node in a linked list '''
    def __init__(self, data=None):
        self.next = None
        self.data = data

    def __str__(self):
        return str(self.data)

class LinkedList:
    ''' simple linked list implementation '''

    def __init__(self):
        self._head = None

    def append(self, data):
        if not self._head:
            self._head = Node(data)
        else:
            node = self._head

            while node and node.next:
                node = node.next

            node.next = Node(data)

    def getHead(self):
        return self._head

    def getNode(self, data):
        node = self._head
        found = None

        if (node is None):
            return

        if node.data is data:
            return node

        while node and node.next:
            if node.data is data:
                found = node
                break

            node = node.next

        return found

    def remove(self, data):
        node = self._head

        if (node is None):
            return

        if node.data is data:
            self._head = node.next
            return

        node = self._head.next
        prev = self._head

        while node and node.next:
            if node.data is data:
                prev.next = node.next
                break

            prev = node
            node = node.next

########################################################

l = LinkedList()

l.append(0)
l.append(1)
l.append(2)
l.append(3)
l.append(4)
l.append(5)
l.append(6)

print(l.getHead()) # 0

l.remove(0)

print(l.getHead()) # 1

l.remove(5)

print(l.getNode(4).next) # 6

